### DISCUSSION TITLE
PyTorch problems facing

### DISCUSSION THREAD

**Discussion by AnirudhG07** [community/community#81383] [Programming Help]
### MNIST accuracy problems

Hello, I need help in PyTorch. I am a beginner and learning PyTorch. I tried making MNIST model. 
SO here is my problem. my model when I ran gave 97% accuracy with test() function.<br>
but when I check manually, I find out the first 100 testing dataset prediction. I am getting 13/100 correct. WHY is this happening??
<img width="1083" alt="image" src="https://github.com/community/community/assets/146579014/6ea2ddad-0058-4fe6-8cdb-76e32ebcb930">

This is my manual check code where I got 13/100 correct.

<img width="1083" alt="image" src="https://github.com/community/community/assets/146579014/baf02b4f-0cd0-4dda-adea-610a43b2653f">
This is my train and test function.

<img width="1083" alt="image" src="https://github.com/community/community/assets/146579014/922a8b31-ac58-4cf3-9226-31fc7b5b956e">
This is my final accuracy of trained code. PLS HELP ME!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by LuluW8071** [comment]
Seems like you are training your model based on Multi-class classification 
If so,

### Install `torchmetrics` for accuracy
```bash
pip install torchmetrics
````

```python
from torchmetrics import Accuracy

accuracy = Accuracy(task = 'multiclass', num_classes = 10).to(device)     
# Change num_classes to your class_labels if it is not 10, you can also do `len(class_labels)`
# setup your loss fn and optimizer ...
````
### Train and Test function
```python
def train_step(model, data_loader, loss_fn, optimizer, accuracy, device):
  train_loss, train_acc = 0, 0
  model.to(device)
  for batch, (X, y) in enumerate(data_loader):
    X, y = X.to(device), y.to(device)
    y_pred = model(X)
    loss = loss_fn(y_pred, y)

    train_loss += loss
    train_acc += accuracy(y, 
        y_pred.argmax(dim = 1)) * 100    # Also you did pred.argmax(0) this selects max value index along columns instead of rows

    # You reversed this part 
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

  train_loss /= len(data_loader)
  train_acc /= len(data_loader)
  print(f"Train loss: {train_loss:.4f} --- Train acc: {train_acc:.2f}%")

def test_step(model, data_loader, loss_fn, optimizer, accuracy, device):
  test_loss, test_acc = 0, 0
  model.to(device)
  model.eval()

  with torch.inference_mode():
    for X_test, y_test in data_loader:
      X_test, y_test = X_test.to(device), y_test.to(device)
      test_pred = model(X_test)

      test_loss += loss_fn(test_pred, y_test)
      test_acc += accuracy(y_test, test_pred.argmax(dim = 1)) * 100 

    test_loss /= len(data_loader)
    test_acc /= len(data_loader)
    print(f'Test loss: {test_loss:.4f} --- Test acc: {test_acc:.2f}%')
```

The recommended order is to 
- first zero the gradients (`optimizer.zero_grad()`), then 
- compute the gradients (`loss.backward()`), and 
- finally update the model parameters (`optimizer.step()`). 

Reversing this order may result in old gradients influencing the new ones, potentially leading to incorrect updates.

Now if your prediction code is correct it should solve the issue else there is some problem on _**model_class**_ or _prediction code snippet_

### Train/Test Loop
```python 
torch.manual_seed(42)   # For reproducibility

# Train/test loop
epochs = 10

for epoch in tqdm(range(epochs)):
  print(f'Epoch: {epoch}')
  train_step(model,
             train_dataloader,
             loss_fn, optimizer,
             accuracy, device)
  test_step(model,
             test_dataloader,
             loss_fn, optimizer,
             accuracy, device)

```
